File: src/llm/enrichSource.ts
/** Default max chars of `source.md` body sent to the enrich LLM (after frontmatter strip). */
export const DEFAULT_ENRICH_MAX_CHARS = 12_000;

const SEP =
  '\n\n---\n*(Đoạn giữa đã lược bỏ để vừa giới hạn ngữ cảnh; ưu tiên phần đầu và phần cuối bài — thường chứa dẫn nhập và kết luận.)*\n---\n\n';

export function enrichMaxCharsFromEnv(): number {
  const n = Number(process.env.ENRICH_MAX_CHARS);
  if (Number.isFinite(n) && n >= 4_000 && n <= 200_000) return Math.floor(n);
  return DEFAULT_ENRICH_MAX_CHARS;
}

/**
 * If body exceeds `maxChars`, keep a head + tail window so conclusions/end of transcript stay in context.
 */
export function truncateSourceForEnrich(body: string, maxChars: number): string {
  if (body.length <= maxChars) return body;
  const budget = maxChars - SEP.length;
  if (budget < 500) return body.slice(0, maxChars);
  const head = Math.floor(budget * 0.62);
  const tail = budget - head;
  return body.slice(0, head) + SEP + body.slice(-tail);
}

File: api/src/brain_api/llm/enrich_source.py
"""Truncate source body for LLM context (mirror cli/src/llm/enrichSource.ts)."""

from __future__ import annotations

import os

DEFAULT_ENRICH_MAX_CHARS = 12_000

_SEP = (
    "\n\n---\n*(Đoạn giữa đã lược bỏ để vừa giới hạn ngữ cảnh; ưu tiên phần đầu và phần cuối bài "
    "— thường chứa dẫn nhập và kết luận.)*\n---\n\n"
)


def enrich_max_chars_from_env() -> int:
    raw = os.environ.get("ENRICH_MAX_CHARS")
    try:
        n = int(raw) if raw else 0
    except ValueError:
        n = 0
    if 4000 <= n <= 200_000:
        return n
    return DEFAULT_ENRICH_MAX_CHARS


def truncate_source_for_enrich(body: str, max_chars: int) -> str:
    if len(body) <= max_chars:
        return body
    budget = max_chars - len(_SEP)
    if budget < 500:
        return body[:max_chars]
    head = int(budget * 0.62)
    tail = budget - head
    return body[:head] + _SEP + body[-tail:]
